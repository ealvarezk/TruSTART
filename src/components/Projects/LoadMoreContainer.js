import styled from "styled-components";

export default styled.div`
  text-align: center;
  margin: 12px;
  height: 32px;
  line-height: 32px;
`;
