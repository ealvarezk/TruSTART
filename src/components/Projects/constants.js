export const PAGE_SIZE = 6;
export const TRUSTAR_LOGO = "https://media-exp1.licdn.com/dms/image/C4E0BAQGsyPwlRDjXUg/company-logo_200_200/0?e=1600300800&v=beta&t=faGsSctG-0R0Fw35600z88FBcg486xF2Dv9YGZJZoek";