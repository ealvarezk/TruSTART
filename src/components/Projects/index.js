import React, { useState } from "react";
import produce from "immer";
import { saveAs } from "file-saver";
import { PageHeader, Alert, Row, Col, List, Button } from "antd";
import { DeleteOutlined, DownloadOutlined } from "@ant-design/icons";
import Project from "../Project";
import useFetchTechport from "../../hooks/useFetchTechport";
import LoadMoreContainer from "./LoadMoreContainer";
import { PAGE_SIZE, TRUSTAR_LOGO } from "./constants";

function Projects() {
  const [{ data, error, loading }, setData] = useFetchTechport();
  const [projectsRendered, setProjectsRendered] = useState(PAGE_SIZE);

  const handleLoadMore = () => {
    setProjectsRendered(projectsRendered + PAGE_SIZE);
  };

  const handleProjectLoaded = (itemId, projectData) => {
    setData(
      produce(data, (draftData) => {
        const project = draftData.projects.projects.find(
          (item) => item.id === itemId
        );
        project.data = projectData;
      })
    );
  };

  const handleSelected = (itemId, selected) => {
    setData(
      produce(data, (draftData) => {
        const project = draftData.projects.projects.find(
          (item) => item.id === itemId
        );
        project.selected = selected;
      })
    );
  };

  const handleDelete = () => {
    setData(
      produce(data, (draftData) => {
        draftData.projects.projects = draftData.projects.projects.filter(
          (item) => !item.selected
        );
      })
    );
  };

  const handleExport = () => {
    const projectsToExport = data.projects.projects.filter(
      (item) => item.selected
    );
    const blob = new Blob([JSON.stringify(projectsToExport)], {
      type: "text/plain;charset=utf-8",
    });
    saveAs(blob, "projects.txt");
  };

  const loadMore = (
    <LoadMoreContainer>
      <Button onClick={handleLoadMore}>Load more...</Button>
    </LoadMoreContainer>
  );

  return (
    <PageHeader
      title="TruSTAR"
      extra={[
        <Button
          key="delete"
          type="primary"
          shape="circle"
          icon={<DeleteOutlined />}
          onClick={handleDelete}
        />,
        <Button
          key="download"
          type="primary"
          shape="circle"
          icon={<DownloadOutlined />}
          onClick={handleExport}
        />,
      ]}
      avatar={{
        src: TRUSTAR_LOGO,
      }}
    >
      <Row style={{ marginTop: 10 }}>
        <Col span={22} offset={1}>
          {error && (
            <Alert
              message="Error Retrieving Projects"
              description={error.message}
              type="error"
              showIcon
              closable
            />
          )}
          {data && (
            <List
              loading={loading}
              loadMore={loadMore}
              dataSource={data.projects.projects.slice(0, projectsRendered)}
              renderItem={(item) => (
                <Project
                  key={item.id}
                  item={item}
                  onLoaded={handleProjectLoaded}
                  onSelected={handleSelected}
                  selected={item.selected}
                />
              )}
            />
          )}
        </Col>
      </Row>
    </PageHeader>
  );
}

export default Projects;
