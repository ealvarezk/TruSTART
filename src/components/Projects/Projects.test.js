import React from "react";
import { shallow } from "enzyme";
import { List, PageHeader } from "antd";
import mockAxios from "jest-mock-axios";
import { saveAs } from "file-saver";

import Projects from "./Projects";

jest.mock('file-saver', () => ({
  saveAs: jest.fn(),
}));

describe("Projects", () => {
  afterEach(() => {
    mockAxios.reset();
  });

  const shallowProjects = () => shallow(<Projects />);

  const shallowProject = (projects) => {
    mockAxios.mockResponse(projectsResponse);

    let list = projects.find(List);
    return shallow(
      <div>
        {list.prop("renderItem")(projectsResponse.data.projects.projects[0])}
      </div>
    ).children();
  };

  const projectsResponse = {
    data: {
      projects: {
        totalCount: 12540,
        projects: [
          { id: 32946, lastUpdated: "2020-6-9" },
          { id: 94227, lastUpdated: "2020-6-5" },
          { id: 11692, lastUpdated: "2020-6-5" },
          { id: 96438, lastUpdated: "2020-6-4" },
          { id: 96439, lastUpdated: "2020-6-4" },
          { id: 94205, lastUpdated: "2020-6-2" },
          { id: 94183, lastUpdated: "2020-6-2" },
          { id: 94884, lastUpdated: "2020-6-1" },
          { id: 13543, lastUpdated: "2020-6-1" },
          { id: 94234, lastUpdated: "2020-5-29" },
          { id: 94233, lastUpdated: "2020-5-29" },
          { id: 94218, lastUpdated: "2020-5-29" },
          { id: 94236, lastUpdated: "2020-5-29" },
          { id: 94240, lastUpdated: "2020-5-21" },
          { id: 10886, lastUpdated: "2020-5-21" },
          { id: 10860, lastUpdated: "2020-5-21" },
        ],
      },
    },
  };

  test("renders Projects properly", () => {
    const projects = shallowProjects();
    
    expect(projects).toMatchSnapshot();
  });

  test("renders Projects properly when error", () => {
    const projects = shallowProjects();
    mockAxios.mockError({ message: "Big Error" });
    expect(projects).toMatchSnapshot();
  });

  test("renders Projects properly when succeed", () => {
    const projects = shallowProjects();
    mockAxios.mockResponse(projectsResponse);
    expect(projects).toMatchSnapshot();
  });

  test("load more click should show more projects", () => {
    const projects = shallowProjects();
    mockAxios.mockResponse(projectsResponse);

    let list = projects.find(List);
    const loadMore = shallow(list.prop("loadMore")).children();
    loadMore.simulate("click");

    list = projects.find(List);

    expect(list.prop("dataSource")).toMatchSnapshot();
  });

  describe("renderItem", () => {
    test("render Project properly", () => {
      const projects = shallowProjects();
      const project = shallowProject(projects);

      expect(project).toMatchSnapshot();
    });

    test("onLoaded should update dataSource", () => {
      const projects = shallowProjects();
      const project = shallowProject(projects);
      project.prop("onLoaded")(32946, { test: "test" });

      const list = projects.find(List);

      expect(list.prop("dataSource")).toMatchSnapshot();
    });

    test("onSelected should update dataSource", () => {
      const projects = shallowProjects();
      const project = shallowProject(projects);
      project.prop("onSelected")(32946, true);

      const list = projects.find(List);

      expect(list.prop("dataSource")).toMatchSnapshot();
    });
  });

  test("on Delete should update dataSource", () => {
    const projects = shallowProjects();
    const project = shallowProject(projects);
    project.prop("onSelected")(32946, true);

    const pageHeader = projects.find(PageHeader);

    const deleteButton = shallow(
      <div>
        {pageHeader.prop('extra')[0]}
      </div>
    ).children();
    deleteButton.simulate('click');

    const list = projects.find(List);
    expect(list.prop("dataSource")).toMatchSnapshot();
  });

  test("on Download should call saveAs", () => {
    const projects = shallowProjects();
    const project = shallowProject(projects);
    project.prop("onSelected")(32946, true);

    const pageHeader = projects.find(PageHeader);

    const downloadButton = shallow(
      <div>
        {pageHeader.prop('extra')[1]}
      </div>
    ).children();
    
    downloadButton.simulate('click');

    expect(saveAs).toHaveBeenCalled();
  });
});
