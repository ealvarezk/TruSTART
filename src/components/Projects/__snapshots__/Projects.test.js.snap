// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects load more click should show more projects 1`] = `
Array [
  Object {
    "id": 32946,
    "lastUpdated": "2020-6-9",
  },
  Object {
    "id": 94227,
    "lastUpdated": "2020-6-5",
  },
  Object {
    "id": 11692,
    "lastUpdated": "2020-6-5",
  },
  Object {
    "id": 96438,
    "lastUpdated": "2020-6-4",
  },
  Object {
    "id": 96439,
    "lastUpdated": "2020-6-4",
  },
  Object {
    "id": 94205,
    "lastUpdated": "2020-6-2",
  },
  Object {
    "id": 94183,
    "lastUpdated": "2020-6-2",
  },
  Object {
    "id": 94884,
    "lastUpdated": "2020-6-1",
  },
  Object {
    "id": 13543,
    "lastUpdated": "2020-6-1",
  },
  Object {
    "id": 94234,
    "lastUpdated": "2020-5-29",
  },
  Object {
    "id": 94233,
    "lastUpdated": "2020-5-29",
  },
  Object {
    "id": 94218,
    "lastUpdated": "2020-5-29",
  },
]
`;

exports[`Projects on Delete should update dataSource 1`] = `
Array [
  Object {
    "id": 94227,
    "lastUpdated": "2020-6-5",
  },
  Object {
    "id": 11692,
    "lastUpdated": "2020-6-5",
  },
  Object {
    "id": 96438,
    "lastUpdated": "2020-6-4",
  },
  Object {
    "id": 96439,
    "lastUpdated": "2020-6-4",
  },
  Object {
    "id": 94205,
    "lastUpdated": "2020-6-2",
  },
  Object {
    "id": 94183,
    "lastUpdated": "2020-6-2",
  },
]
`;

exports[`Projects renderItem onLoaded should update dataSource 1`] = `
Array [
  Object {
    "data": Object {
      "test": "test",
    },
    "id": 32946,
    "lastUpdated": "2020-6-9",
  },
  Object {
    "id": 94227,
    "lastUpdated": "2020-6-5",
  },
  Object {
    "id": 11692,
    "lastUpdated": "2020-6-5",
  },
  Object {
    "id": 96438,
    "lastUpdated": "2020-6-4",
  },
  Object {
    "id": 96439,
    "lastUpdated": "2020-6-4",
  },
  Object {
    "id": 94205,
    "lastUpdated": "2020-6-2",
  },
]
`;

exports[`Projects renderItem onSelected should update dataSource 1`] = `
Array [
  Object {
    "id": 32946,
    "lastUpdated": "2020-6-9",
    "selected": true,
  },
  Object {
    "id": 94227,
    "lastUpdated": "2020-6-5",
  },
  Object {
    "id": 11692,
    "lastUpdated": "2020-6-5",
  },
  Object {
    "id": 96438,
    "lastUpdated": "2020-6-4",
  },
  Object {
    "id": 96439,
    "lastUpdated": "2020-6-4",
  },
  Object {
    "id": 94205,
    "lastUpdated": "2020-6-2",
  },
]
`;

exports[`Projects renderItem render Project properly 1`] = `
<Memo(Project)
  item={
    Object {
      "id": 32946,
      "lastUpdated": "2020-6-9",
    }
  }
  key="32946"
  onLoaded={[Function]}
  onSelected={[Function]}
/>
`;

exports[`Projects renders Projects properly 1`] = `
<PageHeader
  avatar={
    Object {
      "src": "https://media-exp1.licdn.com/dms/image/C4E0BAQGsyPwlRDjXUg/company-logo_200_200/0?e=1600300800&v=beta&t=faGsSctG-0R0Fw35600z88FBcg486xF2Dv9YGZJZoek",
    }
  }
  extra={
    Array [
      <ForwardRef(InternalButton)
        block={false}
        ghost={false}
        htmlType="button"
        icon={<ForwardRef(DeleteOutlined) />}
        loading={false}
        onClick={[Function]}
        shape="circle"
        type="primary"
      />,
      <ForwardRef(InternalButton)
        block={false}
        ghost={false}
        htmlType="button"
        icon={<ForwardRef(DownloadOutlined) />}
        loading={false}
        onClick={[Function]}
        shape="circle"
        type="primary"
      />,
    ]
  }
  title="TruSTAR"
>
  <Row
    gutter={0}
    style={
      Object {
        "marginTop": 10,
      }
    }
  >
    <Col
      offset={1}
      span={22}
    >
      <List
        loadMore={
          <ForwardRef(styled.div)>
            <ForwardRef(InternalButton)
              block={false}
              ghost={false}
              htmlType="button"
              loading={false}
              onClick={[Function]}
            >
              Load more...
            </ForwardRef(InternalButton)>
          </ForwardRef(styled.div)>
        }
        loading={true}
        renderItem={[Function]}
      />
    </Col>
  </Row>
</PageHeader>
`;

exports[`Projects renders Projects properly when error 1`] = `
<PageHeader
  avatar={
    Object {
      "src": "https://media-exp1.licdn.com/dms/image/C4E0BAQGsyPwlRDjXUg/company-logo_200_200/0?e=1600300800&v=beta&t=faGsSctG-0R0Fw35600z88FBcg486xF2Dv9YGZJZoek",
    }
  }
  extra={
    Array [
      <ForwardRef(InternalButton)
        block={false}
        ghost={false}
        htmlType="button"
        icon={<ForwardRef(DeleteOutlined) />}
        loading={false}
        onClick={[Function]}
        shape="circle"
        type="primary"
      />,
      <ForwardRef(InternalButton)
        block={false}
        ghost={false}
        htmlType="button"
        icon={<ForwardRef(DownloadOutlined) />}
        loading={false}
        onClick={[Function]}
        shape="circle"
        type="primary"
      />,
    ]
  }
  title="TruSTAR"
>
  <Row
    gutter={0}
    style={
      Object {
        "marginTop": 10,
      }
    }
  >
    <Col
      offset={1}
      span={22}
    >
      <Alert
        closable={true}
        description="Big Error"
        message="Error Retrieving Projects"
        showIcon={true}
        type="error"
      />
    </Col>
  </Row>
</PageHeader>
`;

exports[`Projects renders Projects properly when succeed 1`] = `
<PageHeader
  avatar={
    Object {
      "src": "https://media-exp1.licdn.com/dms/image/C4E0BAQGsyPwlRDjXUg/company-logo_200_200/0?e=1600300800&v=beta&t=faGsSctG-0R0Fw35600z88FBcg486xF2Dv9YGZJZoek",
    }
  }
  extra={
    Array [
      <ForwardRef(InternalButton)
        block={false}
        ghost={false}
        htmlType="button"
        icon={<ForwardRef(DeleteOutlined) />}
        loading={false}
        onClick={[Function]}
        shape="circle"
        type="primary"
      />,
      <ForwardRef(InternalButton)
        block={false}
        ghost={false}
        htmlType="button"
        icon={<ForwardRef(DownloadOutlined) />}
        loading={false}
        onClick={[Function]}
        shape="circle"
        type="primary"
      />,
    ]
  }
  title="TruSTAR"
>
  <Row
    gutter={0}
    style={
      Object {
        "marginTop": 10,
      }
    }
  >
    <Col
      offset={1}
      span={22}
    >
      <List
        dataSource={
          Array [
            Object {
              "id": 32946,
              "lastUpdated": "2020-6-9",
            },
            Object {
              "id": 94227,
              "lastUpdated": "2020-6-5",
            },
            Object {
              "id": 11692,
              "lastUpdated": "2020-6-5",
            },
            Object {
              "id": 96438,
              "lastUpdated": "2020-6-4",
            },
            Object {
              "id": 96439,
              "lastUpdated": "2020-6-4",
            },
            Object {
              "id": 94205,
              "lastUpdated": "2020-6-2",
            },
          ]
        }
        loadMore={
          <ForwardRef(styled.div)>
            <ForwardRef(InternalButton)
              block={false}
              ghost={false}
              htmlType="button"
              loading={false}
              onClick={[Function]}
            >
              Load more...
            </ForwardRef(InternalButton)>
          </ForwardRef(styled.div)>
        }
        loading={false}
        renderItem={[Function]}
      />
    </Col>
  </Row>
</PageHeader>
`;
